cmd/gpu_nfdhook/labeler.go

```diff
@@ -171,33 +171,70 @@ func (l *labeler) createCapabilityLabels(cardNum string, numTiles uint64) {
 	}
 	defer file.Close()
 
-	// define strings to search from the file, and the actions to take in order to create labels from those strings (as funcs)
+	gen := ""
+	media := ""
+	graphics := ""
+	// define string prefixes to search from the file, and the actions to take in order to create labels from those strings (as funcs)
 	searchStringActionMap := map[string]func(string){
-		"platform: ": func(platformName string) {
+		"platform:": func(platformName string) {
 			l.labels.addNumericLabel(labelNamespace+"platform_"+platformName+".count", 1)
 			l.labels[labelNamespace+"platform_"+platformName+".tiles"] = strconv.FormatInt(int64(numTiles), 10)
 			l.labels[labelNamespace+"platform_"+platformName+".present"] = "true"
 		},
-		"gen: ": func(genName string) {
-			l.labels[labelNamespace+"platform_gen"] = genName
+		// there's also display block version, but that's not relevant
+		"media version:": func(version string) {
+			l.labels[labelNamespace+"media_version"] = version
+			media = version
+		},
+		"graphics version:": func(version string) {
+			l.labels[labelNamespace+"graphics_version"] = version
+			graphics = version
+		},
+		"gen:": func(version string) {
+			l.labels[labelNamespace+"platform_gen"] = version
+			gen = version
 		},
 	}
 
 	// Finally, read the file, and try to find the matches. Perform actions and reduce the search map size as we proceed. Return at 0 size.
 	scanner := bufio.NewScanner(file)
+scanning:
 	for scanner.Scan() {
-		for searchString, action := range searchStringActionMap {
-			var stringValue string
-			n, _ := fmt.Sscanf(scanner.Text(), searchString+"%s", &stringValue)
-			if n > 0 {
-				action(stringValue)
-				delete(searchStringActionMap, searchString)
-				if len(searchStringActionMap) == 0 {
-					return
-				}
-				break
+		line := scanner.Text()
+		for prefix, action := range searchStringActionMap {
+			if !strings.HasPrefix(line, prefix) {
+				continue
+			}
+			fields := strings.Split(line, ": ")
+			if len(fields) == 2 {
+				action(fields[1])
+			} else {
+				klog.Warningf("invalid '%s' line format: '%s'", file.Name(), line)
+			}
+			delete(searchStringActionMap, prefix)
+			if len(searchStringActionMap) == 0 {
+				break scanning
 			}
```

Comment:
I guess now this check can be moved right after the loop. Then there would be no need for the loop label.

Reply:
Then it would be run for every capabilities line, instead of only when one of the prefixes matches.  With drm-tip on DG1:
```
# wc -l /sys/kernel/debug/dri/*/i915_capabilities 
104 /sys/kernel/debug/dri/0/i915_capabilities
```
i.e. 104 instead of 2-3 times.

Reply:
It keeps iterating because it never becomes empty with current kernels, and while both "gen" and *_version fields are supported.  All the kernel capability files I have seen, have either "gen" or *_version fields, not both.

Reply:
I missed something. What never becomes empty? `searchStringActionMap`?

If one of the prefixes matches then the internal loop breaks and doesn't iterate regardless of the map's emptiness.

Reply:
Yes, currently `searchStringActionMap` never becomes empty for the reason I stated, therefore all lines in the capability file will be parsed (even before this PR, in case of newer kernels).  Along with the number of map keys increasing, that's why this PR optimizes the parsing.

Reply:
Ok. Now I see what you mean. Thanks! Let's keep it this way too.